File: src/ros2_core/tf2_pkg/src/tf2_listener.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/exceptions.h"

using namespace std::chrono_literals;

class TF2ListenerNode : public rclcpp::Node
{
    public:
        TF2ListenerNode() : Node("tf2_listener_node")
        {
            RCLCPP_INFO(this->get_logger(), "%s has been started!", this->get_name());

            // 1. Create the Buffer (The Memory Cache)
            // In C++, the Buffer explicitly requires the node's clock to track time accurately.
            tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());

            // 2. Create the Listener (The Network Subscriber)
            // We pass a reference to the buffer so the listener can feed incoming data into it.
            tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

            // 3. Create a timer to query the buffer every 1 second (1.0 Hz)
            timer_ = this->create_timer(
                1000ms,
                [this]() -> void { this->timer_callback(); }
            );
        }

    private:
        std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
        std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
        rclcpp::TimerBase::SharedPtr timer_;

        void timer_callback()
        {
            std::string target_frame = "world";
            std::string source_frame = "my_dynamic_frame";

            try {
                // 4. Query the Buffer
                // tf2::TimePointZero is the C++ equivalent of asking for the "newest available transform"
                geometry_msgs::msg::TransformStamped t = tf_buffer_->lookupTransform(
                    target_frame,
                    source_frame,
                    tf2::TimePointZero
                );

                // Extract the coordinates and print them!
                double x = t.transform.translation.x;
                double y = t.transform.translation.y;
                double z = t.transform.translation.z;

                RCLCPP_INFO(this->get_logger(), "Dynamic Frame Location -> X: %.2f, Y: %.2f, Z: %.2f", x, y, z);
            }
            catch (const tf2::TransformException & ex) {
                // If the frames don't exist yet, or aren't connected, lookupTransform throws an exception.
                // We MUST catch it so our node doesn't crash!
                RCLCPP_WARN(
                    this->get_logger(), 
                    "Could not transform %s to %s: %s", 
                    target_frame.c_str(), 
                    source_frame.c_str(), 
                    ex.what()
                );
            }
        }
};

int main(int argc, char * argv[])
{
    auto log = rclcpp::get_logger("System");
    TF2ListenerNode::SharedPtr node_instance = nullptr;

    try {
        RCLCPP_INFO(log, "Initializing the ROS2 Client...");
        rclcpp::init(argc, argv);

        RCLCPP_INFO(log, "Starting a ROS2 Node...");
        node_instance = std::make_shared<TF2ListenerNode>();
        rclcpp::spin(node_instance);

        RCLCPP_WARN(log, "[CTRL+C]>>> Interrupted by the User.");
        RCLCPP_INFO(log, "Destroying the ROS2 Node...");
    }
    catch (const std::exception & e) {
        RCLCPP_ERROR(log, "Critical Error: %s", e.what());
    }

    if (rclcpp::ok()) {
        RCLCPP_INFO(log, "Manually shutting down the ROS2 Client...");
        rclcpp::shutdown();
    }
    return 0;
}

File: src/ros2_core/tf2_pkg/tf2_pkg/tf2_listener.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
from rclpy.time import Time
from rclpy.clock import ClockType

# TF2 specific imports
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from tf2_ros import TransformException  # type: ignore

class TF2ListenerNode(Node):
    """
    This node listens to the TF2 tree and queries the spatial relationship
    between two frames using a Buffer and a Listener.
    """
    def __init__(self):
        super().__init__('tf2_listener_node')
        self.get_logger().info(f"{self.get_name()} has been started!")

        # 1. Create the Buffer (The Memory Cache)
        # It stores up to 10 seconds of transform history by default.
        self.tf_buffer = Buffer()

        # 2. Create the Listener (The Network Subscriber)
        # This automatically hooks into /tf and /tf_static and feeds the Buffer.
        self.tf_listener = TransformListener(self.tf_buffer, self)

        # 3. Create a timer to query the buffer every 1 second (1.0 Hz)
        self.timer = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        # We want to know: "Where is my_dynamic_frame relative to the world?"
        target_frame = 'world'
        source_frame = 'my_dynamic_frame'

        try:
            # 4. Query the Buffer
            # Time() means "give me the newest available transform"
            t = self.tf_buffer.lookup_transform(
                target_frame,
                source_frame,
                Time(clock_type=ClockType.ROS_TIME) 
            )

            # Extract the coordinates and print them!
            x = t.transform.translation.x
            y = t.transform.translation.y
            z = t.transform.translation.z
            
            self.get_logger().info(f"Dynamic Frame Location -> X: {x:.2f}, Y: {y:.2f}, Z: {z:.2f}")

        except TransformException as ex:
            # If the frames don't exist yet, or aren't connected, lookup_transform throws an exception.
            # We MUST catch it so our node doesn't crash!
            self.get_logger().warn(f"Could not transform {target_frame} to {source_frame}: {ex}")

def main(args=None) -> None:
    log = get_logger("System")
    node_instance = None

    try:
        log.info("Initializing the ROS2 Client...")
        rclpy.init(args=args)

        log.info("Starting a ROS2 Node...")
        node_instance = TF2ListenerNode()
        
        rclpy.spin(node_instance)

    except KeyboardInterrupt:
        log.warn("[CTRL+C]>>> Interrupted by the User.")

    except Exception as e:
        log.error(f"Critical Error: {e}")
    
    finally:
        if node_instance is not None:
            log.info("Destroying the ROS2 Node...")
            node_instance.destroy_node()
            node_instance = None

        if rclpy.ok():
            log.info("Manually shutting down the ROS2 Client...")
            rclpy.shutdown()

if __name__ == '__main__':
    main()